src/sequence.rs

```diff
@@ -67,35 +104,73 @@ where
     /// Get the current position in the sequence.
     #[inline]
     pub fn index(&self) -> T {
-        self.current_index.wrapping_sub(&self.start_index)
+        self.current_index
     }
 }
 
-impl<T> Iterator for RandomSequence<T>
-where
-    T: PrimInt + WrappingAdd + WrappingSub + QuadraticResidue
-{
-    type Item = T;
+macro_rules! impl_unsized_iterator {
+    ($T:ident) => {
+        impl Iterator for RandomSequence<$T> {
+            type Item = $T;
 
-    #[inline]
-    fn next(&mut self) -> Option<Self::Item> {
-        Some(self.next())
-    }
+            #[inline]
+            fn next(&mut self) -> Option<Self::Item> {
+                self.next()
+            }
+
+            #[inline]
+            fn size_hint(&self) -> (usize, Option<usize>) {
+                ($T::MAX as usize, None)
+            }
+        }
+    };
 }
 
+macro_rules! impl_exact_size_iterator {
+    ($T:ident) => {
+        impl Iterator for RandomSequence<$T> {
+            type Item = $T;
+
+            #[inline]
+            fn next(&mut self) -> Option<Self::Item> {
+                self.next()
+            }
+
+            #[inline]
+            fn size_hint(&self) -> (usize, Option<usize>) {
+                ($T::MAX as usize + 1, Some($T::MAX as usize + 1))
+            }
+        }
+
+        impl ExactSizeIterator for RandomSequence<$T> {}
+    };
+}
+
+// Can only fit exact size iterators in types smaller than usize. As usize will have usize+1 elements.
+impl_exact_size_iterator!(u8);
+impl_exact_size_iterator!(u16);
+#[cfg(target_pointer_width = "64")]
+impl_exact_size_iterator!(u32);
+#[cfg(target_pointer_width = "32")]
+impl_unsized_iterator!(u32);
+impl_unsized_iterator!(u64);
+impl_unsized_iterator!(usize);
```

Comment:
Awkward macros to avoid `impl ExactSizeIterator` for types where the sequence length would overflow `usize`.